### DISCUSSION TITLE
How to bold filename in default status bar?

### DISCUSSION THREAD

**Discussion by pkazmier** [LazyVim/LazyVim#2605] [Q&A] [answered]
How can I bold _only_ the filename in the path displayed in LazyVim's default status line? I think it would really improve readability and emphasize the most important aspect of the path.

The last component of section C is `Util.lualine.pretty_path()`, which shortens the path intelligently and changes the color of the filename portion if it has been modified. I've looked at this function, but cannot figure out how to always bold the filename portion of the path—while still keeping color changing when the buffer is modified.

Would appreciate any help. Thank you!

**Comment by dpetka2001** [comment]
Not without tampering with the function itself. I would suggest just copying the functions over to your personal configuration and use those instead, because if you override the functions themselves in the Lazyvim config they might be overriden in a future update and it would break your configuration. You could do something like
```lua
-- ~/.config/nvim/lua/plugins/lualine.lua
local Util = require("lazyvim.util")
local function format(component, text, hl_group)
  if not hl_group then
    return text
  end
  ---@type table<string, string>
  component.hl_cache = component.hl_cache or {}
  local lualine_hl_group = component.hl_cache[hl_group]
  if not lualine_hl_group then
    local utils = require("lualine.utils.utils")
    local mygui = function()
      local mybold = utils.extract_highlight_colors(hl_group, "bold") and "bold"
      local myitalic = utils.extract_highlight_colors(hl_group, "italic") and "italic"
      if mybold and myitalic then
        return mybold .. "," .. myitalic
      elseif mybold then
        return mybold
      elseif myitalic then
        return myitalic
      else
        return ""
      end
    end

    lualine_hl_group = component:create_hl({
      fg = utils.extract_highlight_colors(hl_group, "fg"),
      gui = mygui(),
    }, "LV_" .. hl_group)
    component.hl_cache[hl_group] = lualine_hl_group
  end
  return component:format_hl(lualine_hl_group) .. text .. component:get_default_hl()
end

---@param opts? {relative: "cwd"|"root", modified_hl: string?}
local function pretty_path(opts)
  opts = vim.tbl_extend("force", {
    relative = "cwd",
    modified_hl = "Constant",
    filename_hl = "",
  }, opts or {})

  return function(self)
    local path = vim.fn.expand("%:p") --[[@as string]]

    if path == "" then
      return ""
    end
    local root = Util.root.get({ normalize = true })
    local cwd = Util.root.cwd()

    if opts.relative == "cwd" and path:find(cwd, 1, true) == 1 then
      path = path:sub(#cwd + 2)
    else
      path = path:sub(#root + 2)
    end

    local sep = package.config:sub(1, 1)
    local parts = vim.split(path, "[\\/]")
    if #parts > 3 then
      parts = { parts[1], "…", parts[#parts - 1], parts[#parts] }
    end

    if opts.modified_hl and vim.bo.modified then
      parts[#parts] = format(self, parts[#parts], opts.modified_hl)
    else
      parts[#parts] = format(self, parts[#parts], opts.filename_hl)
    end

    return table.concat(parts, sep)
  end
end

return {
  {
    "nvim-lualine/lualine.nvim",
    opts = function(_, opts)
      opts.sections.lualine_c[4] = { pretty_path({ filename_hl = "MyErrorMsg" }) }
    end,
  },
}
```
and you would also need to define a custom highlight group in a autocmd for the highlight group, something like
```lua
-- ~/.config/nvim/lua/config/autocmds.lua
vim.api.nvim_create_autocmd("BufReadPre", {
  group = vim.api.nvim_create_augroup("MyErrorMsg", { clear = true }),
  desc = "MyErrorMsg",
  command = "hi MyErrorMsg gui=bold",
})
```
I tested with the above configuration and it worked like you want I believe.

**Reply by pkazmier** [reply]
Thank you very much for the detailed response. I will test later tonight.

**Reply by pkazmier** [reply, marked as answer]
This worked well by the way. Thank you so much.

I've also tweaked the code further to deemphasize the directory path to reduce the visual clutter without removing it entirely. I think it looks so much better than LazyVim’s default. Here are some screenshots to illustrate:

Lazy's default:
<img width="929" alt="image" src="https://github.com/LazyVim/LazyVim/assets/747855/2c9e2680-5700-43e3-bd7c-8c104b152337">

Our combined efforts:
<img width="930" alt="image" src="https://github.com/LazyVim/LazyVim/assets/747855/d08f22bb-5fc4-4ec9-8f98-954ed604abfe">

When the buffer has been modified:
<img width="929" alt="image" src="https://github.com/LazyVim/LazyVim/assets/747855/7eb16f47-a7d5-44bb-bb7b-a52a277cb0b2">

Here is my modified `pretty_path`, which uses your `format` from above:

```lua
local function pretty_path(opts)
  opts = vim.tbl_extend("force", {
    relative = "cwd",
    modified_hl = "NeogitGraphBoldYellow",
    filename_hl = "NeogitGraphBoldWhite",
    dirpath_hl = "Conceal",
  }, opts or {})

  return function(self)
    local path = vim.fn.expand("%:p") --[[@as string]]
    if path == "" then
      return ""
    end

    local root = util.root.get({ normalize = true })
    local cwd = util.root.cwd()

    if opts.relative == "cwd" and path:find(cwd, 1, true) == 1 then
      path = path:sub(#cwd + 2)
    else
      path = path:sub(#root + 2)
    end

    local sep = package.config:sub(1, 1)
    local parts = vim.split(path, "[\\/]")

    if #parts > 3 then
      parts = { parts[1], "…", parts[#parts - 1], parts[#parts] }
    end

    if opts.modified_hl and vim.bo.modified then
      parts[#parts] = format(self, parts[#parts], opts.modified_hl)
    else
      parts[#parts] = format(self, parts[#parts], opts.filename_hl)
    end

    local dirpath = ""
    if #parts > 1 then
      dirpath = table.concat({ unpack(parts, 1, #parts - 1) }, sep)
      dirpath = format(self, dirpath .. sep, opts.dirpath_hl)
    end
    return dirpath .. parts[#parts]
  end
end
```

**Reply by dpetka2001** [reply]
Cool!! I really like how it looks with the `Conceal`. Nice job :smiley:

**Reply by karimlevallois** [reply]
Would you mind sharing your dotfiles please, can't seem to get this to work on my side for some reason.

**Reply by pkazmier** [reply]
@karimlevallois Sure.

```lua
-- ~/.config/nvim/lua/plugins/lualine.lua
local util = require("lazyvim.util")

local function format(component, text, hl_group)
  if not hl_group then
    return text
  end

  ---@type table<string, string>
  component.hl_cache = component.hl_cache or {}
  local lualine_hl_group = component.hl_cache[hl_group]
  if not lualine_hl_group then
    local utils = require("lualine.utils.utils")
    local mygui = function()
      local mybold = utils.extract_highlight_colors(hl_group, "bold") and "bold"
      local myitalic = utils.extract_highlight_colors(hl_group, "italic") and "italic"
      if mybold and myitalic then
        return mybold .. "," .. myitalic
      elseif mybold then
        return mybold
      elseif myitalic then
        return myitalic
      else
        return ""
      end
    end

    lualine_hl_group = component:create_hl({
      fg = utils.extract_highlight_colors(hl_group, "fg"),
      gui = mygui(),
    }, "LV_" .. hl_group)
    component.hl_cache[hl_group] = lualine_hl_group
  end
  return component:format_hl(lualine_hl_group) .. text .. component:get_default_hl()
end

local function pretty_path(opts)
  opts = vim.tbl_extend("force", {
    relative = "cwd",
    modified_hl = "NeogitGraphBoldYellow",
    filename_hl = "NeogitGraphBoldWhite",
    dirpath_hl = "Conceal",
  }, opts or {})

  return function(self)
    local path = vim.fn.expand("%:p") --[[@as string]]
    if path == "" then
      return ""
    end

    local root = util.root.get({ normalize = true })
    local cwd = util.root.cwd()

    if opts.relative == "cwd" and path:find(cwd, 1, true) == 1 then
      path = path:sub(#cwd + 2)
    else
      path = path:sub(#root + 2)
    end

    local sep = package.config:sub(1, 1)
    local parts = vim.split(path, "[\\/]")

    if #parts > 3 then
      parts = { parts[1], "…", parts[#parts - 1], parts[#parts] }
    end

    if opts.modified_hl and vim.bo.modified then
      parts[#parts] = format(self, parts[#parts], opts.modified_hl)
    else
      parts[#parts] = format(self, parts[#parts], opts.filename_hl)
    end

    local dirpath = ""
    if #parts > 1 then
      dirpath = table.concat({ unpack(parts, 1, #parts - 1) }, sep)
      dirpath = format(self, dirpath .. sep, opts.dirpath_hl)
    end
    return dirpath .. parts[#parts]
  end
end

return {
  {
    "nvim-lualine/lualine.nvim",
    opts = function(_, opts)
      opts.options = {
        section_separators = { left = "", right = "" },
        component_separators = { left = "", right = "" },
      }
      opts.sections.lualine_a = {
        {
          function()
            return ""
          end,
          padding = { left = 1, right = 0 },
          separator = "",
        },
        "mode",
      }
      opts.sections.lualine_c[4] = { pretty_path( { filename_hl = "Bold", modified_hl = "MatchParen" } ) } 
      opts.sections.lualine_y = { "progress" }
      opts.sections.lualine_z = {
        { "location", separator = "" },
        {
          function()
            return ""
          end,
          padding = { left = 0, right = 1 },
        },
      }
      return opts
    end,
  },
}
```

**Reply by pkazmier** [reply]
@karimlevallois I did a fresh lazy install and placed the above file in my plugins directory and everything worked.

**Reply by karimlevallois** [reply]
It's working beautifully, thank you.

**Reply by pkazmier** [reply]
@dpetka2001 Do you think I should submit a PR for this? I know we're biased, but it really does make the UI look a lot better.

**Reply by dpetka2001** [reply]
I believe this is a stylistic user preference and it should be left to the user how he wants to look like. The functions already exist in LazyVim so no need for that I believe. People who care about this can do it in their personal configuration. Not every user's preference can be taken into account in a Neovim distro.

**Reply by pkazmier** [reply]
I agree with that point, but the existing `pretty_path` function only provides the ability to change the **modified** face. With our changes, users have the _option_ to change the group used for two additional components: directory path and filename. Plus, the existing `format` function didn't support using gui styles, which you had to enhance.

So, a user who wants to achieve the same look, has to install a lot of code to do so vs just doing:

```lua
opts.sections.lualine_c[4] = { 
  pretty_path( { 
    filename_hl = "Bold",
    dirpath_hl = "Conceal",
    modified_hl = "MatchParen",
  } ) } 
```

I guess the way I look at it is that it's only providing additional options.  I can set the default groups in the PR so the behavior is identical to the out-of-box lazyvim behavior today. Then, if a user wants to format the appearance of their path, they don't have to install a bunch of code, but instead just pass options to `pretty_path`.

If the maintainer doesn't like it, he can just reject. No harm in trying I suppose.

**Reply by dpetka2001** [reply]
Yeah sure, go ahead. Just add the extra necessary components and leave the default style as is. Then users can change it if they'd like. Definitely doesn't hurt to try :smile:

**Reply by januz** [reply]
@pkazmier Thanks for submitting this as a PR. I'm trying to get the same style working but am only able to get the filename bolded. The `directory_hl = "Conceal"` and `modified_hl = "MatchParen"` don't seem to be doing anything for me, i.e., the directory name is not less visible and the color of the filename does not change when the file has been modified. Are there additional things that I have to configure besides the `lualine` config to get these to work? Thanks so much!

**Reply by dpetka2001** [reply]
This probably depends on the colorscheme you're using and how the highlight groups are configured for it. In Tokyonight (which is LazyVim's default colorscheme) it works correctly. Maybe you should change the highlight groups used according to your preferences.

**Reply by januz** [reply]
That makes sense. Thanks for pointing me in the right direction!

**Comment by nipuna-perera** [comment]
Would love for this to be a LazyVim default!

**Reply by pkazmier** [reply]
Well, I submitted #2687.  Let's see if the maintainer accepts it. The PR leaves the default as is, so users can override it easily, but I really do think the default for Lazy should change to this as it's nicer looking (IMHO).

**Reply by dpetka2001** [reply]
@pkazmier Just for future reference, I'm just a regular user like you with no affiliation to this project. I've just contributed some small stuff. So, whenever you feel like you have a good idea about improving Lazyvim, just go ahead and make a PR. No need to ask me. The maintainer is the one who makes decisions. I hope i didn't sound discouraging in my previous post. If I did, I apologize.

PS: I really like your PR and hope it gets approved :)

**Reply by folke** [reply]
Made some small changes and merged it in the meantime. Looks great. Thanks!
